test/server.ml

```diff
@@ -1084,6 +1084,103 @@ module A = struct
                 __LOC__ (Ok (Dns_trie.empty, [ n_of_s "one.com", soa'' ]))
                 (handle_update server' `Udp (Some key) q up'))
 
+  let actual_update_hostname_key () =
+    let server = server () in
+    let q = Packet.Question.create (n_of_s "foo.one.com") Soa in
+    let foo, entry_key, entry_val =
+      n_of_s "foo.one.com", Rr_map.A,
+      (300l, Rr_map.Ipv4_set.singleton (ip_of_s "127.0.0.1"))
+    in
+    let up =
+      Domain_name.Map.empty,
+      Domain_name.Map.singleton foo [
+        Packet.Update.Add Rr_map.(B (entry_key, entry_val))
+      ]
+    in
+    let key = n_of_s "mykey._update.foo.one.com" in
+    let soa' = { soa with serial = Int32.succ soa.serial } in
+    let trie' =
+      Dns_trie.insert (n_of_s "one.com") Rr_map.Soa soa'
+        (Dns_trie.insert foo entry_key entry_val example_trie)
+    in
+    Alcotest.(check (result (pair trie_test zone_test) rcode_test)
+                __LOC__ (Ok (trie', [ n_of_s "one.com", soa' ]))
+                (handle_update server `Udp (Some key) q up))
+
+  let trie_with_two trie =
+    Dns_trie.insert (n_of_s "two.com") Soa soa
+      (Dns_trie.insert (n_of_s "two.com") Ns (300l, example_ns) trie)
+
+  let update_zone_regression () =
+    (* implementation incremented serial of all zones for one update *)
+    let server = server () in
+    let server' = Dns_server.with_data server (trie_with_two server.data) in
+    let q = Packet.Question.create (n_of_s "one.com") Soa in
+    let foo, entry_key, entry_val =
+      n_of_s "foo.one.com", Rr_map.A,
+      (300l, Rr_map.Ipv4_set.singleton (ip_of_s "127.0.0.1"))
+    in
+    let up =
+      Domain_name.Map.empty,
+      Domain_name.Map.singleton foo [
+        Packet.Update.Add Rr_map.(B (entry_key, entry_val))
+      ]
+    in
+    let key = n_of_s "foo._update.one.com" in
+    let soa' = { soa with serial = Int32.succ soa.serial } in
+    let trie' =
+      Dns_trie.insert (n_of_s "one.com") Rr_map.Soa soa'
+        (Dns_trie.insert foo entry_key entry_val server'.data)
+    in
+    Alcotest.(check (result (pair trie_test zone_test) rcode_test)
+                __LOC__ (Ok (trie', [ n_of_s "one.com", soa' ]))
+                (handle_update server' `Udp (Some key) q up));
+    let server'' = Dns_server.with_data server' trie' in
```

Comment:
maybe move `server''` bind down below `old_trie` to make it clearer that `old_trie` is not using `server'` by mistake?